### PR TITLE
Some improvements to tensor_block() and put/get_raw()

### PR DESCRIPTION
**Pull request by jcorbino-lbl**

### CODE REVIEW

File: src/tamm/tensor_impl.hpp [hunk 1]
```diff
@@ -1189,8 +1194,9 @@ class DenseTensorImpl: public TensorImpl<T> {
             upcxx::global_ptr<uint8_t> remote_addr =
               gptrs_[t.rank] + (t.offset * elem_sz) + (tile_offset * elem_sz);
             uint8_t* local_addr = ((uint8_t*) buf) + (next++ * elem_sz);
-            upcxx::rget(remote_addr, local_addr, elem_sz).wait();
+            upcxx::rget(remote_addr, local_addr, elem_sz, upcxx::operation_cx::as_promise(p));
           }
+    p.finalize().wait();
```
Comment (bonachea) [inline src/tamm/tensor_impl.hpp:1199]: Similarly here, please do not call fine-grained `upcxx::rget` in a loop like this.
just push the addresses into vectors, and then make an `rget_regular()` call.

**Review by bonachea — CHANGES_REQUESTED** [review]
Added a requested change to use of `upcxx::{rput,rget}_regular()`. This should be a simple local code replacement and could yield orders-of-magnitude performance improvement to the operations in question.

File: src/tamm/tensor_impl.hpp [hunk 2]
```diff
@@ -1154,7 +1135,30 @@ class DenseTensorImpl: public TensorImpl<T> {
...
+            int64_t    k_offset = k - t.lo[2];
+            int64_t    l_offset = l - t.lo[3];
+            int64_t    tile_offset =
+              l_offset + t.dim[3] * (k_offset + t.dim[2] * (j_offset + t.dim[1] * i_offset));
+            upcxx::global_ptr<uint8_t> remote_addr =
+              gptrs_[t.rank] + (t.offset * elem_sz) + (tile_offset * elem_sz);
+            uint8_t* local_addr = ((uint8_t*) buf) + (next++ * elem_sz);
+            upcxx::rput(local_addr, remote_addr, elem_sz, upcxx::operation_cx::as_promise(p));
+          }
+    p.finalize().wait();
```
[the remark this answers is not in the document]
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1158]: I understand the importance of aggregating communication. The thing here is the following:
A tensor block is potentially spread among multiple processes (tiles that are affine to different processes), and the chunks of contiguous elements could have different sizes.  And even via a rput/rget_irregular one will pull some unnecessary elements so it is quite more elaborated (working on that BTW). Also worth mentioning that the only routine that invokes these fine-grained puts/gets is tensor_block which is not a frequent operation. I also forgot to mention in the PR description that this is not the ultimate version of this routines, but certainly better than the current :)

So again, the sizes that one needs to grab could be different but not only that, they're not necessarily contiguous within the tile.

[the remark this answers is not in the document]
Comment (bonachea) [reply src/tamm/tensor_impl.hpp:1158]: @jcorbino-lbl I don't see how anything in your response argues against the substitution I'm suggesting, Every `rput` in the code above is using a fixed element size of `elem_sz` with different addresses, which is exactly the use case for `rput_regular`. It doesn't matter if some of the elements happen to be contiguous, that doesn't affect the operation of`rput_regular` and aggregation will still be MUCH faster than a separate `rput` operation for each element.

The unordered_map approach I outlined handles the case where tiles are spread across different remote ranks, issuing exactly one `rput_regular` per peer.

In short, I really think the difference between the loop you've written that issues one fine-grained  `rput` per innermost iteration, and the one I'm suggesting that instead pushes addresses into a map of vectors and then issues one  `rput_regular` per peer at the end should amount to less than 10 additional lines of code (the 10 lines in my sketch above). It's difficult to overstate the performance benefit those 10 lines of code should provide by enabling aggregation, so I don't see a good reason not to just write it correctly the first time around.

[the remark this answers is not in the document]
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1158]: I think we're not communicating, what I meant by different sizes is that the amount of elements that one needs to get/put per tile is not necessarily the same. I wasn't talking about the size of the type (sorry for the confusion). Correct me if I'm wrong but that rget/rput_regular will get/put contiguous elements for each pair of src/dst, am I correct? So my point is that it is more like a strided one but for each tile involved in the block, no?

[the remark this answers is not in the document]
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1158]: for instance:
`
pair<particle_t*, size_t> src[]={{srcP+12, 22}, {srcP+66, 12}, {srcP, 4}};
pair<upcxx::global_ptr<particle_t>, size_t> dest[]={{destP, 38}};
auto f = upcxx::rput_irregular(src, end(src), dest, end(dest));
`
all the elements in the range srcP+12, 22 are contiguous. Right? So is not only about specifying the arrays with different gptrs but also the amount per pair is different and in some tiles they're not contiguous. Sorry if I still don't get your point.

[the remark this answers is not in the document]
Comment (bonachea) [reply src/tamm/tensor_impl.hpp:1158]: > rget/rput_regular will get/put contiguous elements for each pair of src/dst, am I correct?

Correct. `{rget,rput}_regular()` takes a sequence of element addresses for each end and aggregates them all into one bandwidth-efficient stream of operations. These functions work at the level of individual element addresses (which your loop is already calculating) and don't care whether those addresses are part of the same tile or whether they happen to be contiguous.

Of course later we might come back and improve upon this further using strided operations or `{rget,rput}_irregular()` to further aggregate adjacent elements, but performance-wise that's a minor refinement compared to the difference between no aggregation and use of `{rget,rput}_regular()`.

[the remark this answers is not in the document]
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1158]: You mean simply pushing the addresses in the vector (the address of  each elem) and doing the put/get out of the loop with all the addresses?

[the remark this answers is not in the document]
Comment (bonachea) [reply src/tamm/tensor_impl.hpp:1158]: > for instance: `pair<particle_t*, size_t> src[]={{srcP+12, 22}, {srcP+66, 12}, {srcP, 4}}; pair<upcxx::global_ptr<particle_t>, size_t> dest[]={{destP, 38}}; auto f = upcxx::rput_irregular(src, end(src), dest, end(dest));` all the elements in the range srcP+12, 22 are contiguous. Right? So is not only about specifying the arrays with different gptrs but also the amount per pair is different and in some tiles they're not contiguous. Sorry if I still don't get your point.

You are looking at the example for IRREGULAR `{rget,rput}_irregular()`, which is NOT what I'm suggesting here. I'm suggesting `{rget,rput}_regular()` which is a perfect match to this loop.

[the remark this answers is not in the document]
Comment (bonachea) [reply src/tamm/tensor_impl.hpp:1158]: > You mean simply pushing the addresses in the vector (the address of each elem) and doing the put/get out of the loop with all the addresses?

Exactly.

[the remark this answers is not in the document]
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1158]: I get it know! Will do that!!! Thank you!

File: src/tamm/tensor_impl.hpp [hunk 3]
```diff
@@ -1153,8 +1157,16 @@ class DenseTensorImpl: public TensorImpl<T> {
             upcxx::global_ptr<uint8_t> remote_addr =
               gptrs_[t.rank] + (t.offset * elem_sz) + (tile_offset * elem_sz);
             uint8_t* local_addr = ((uint8_t*) buf) + (next++ * elem_sz);
-            upcxx::rput(local_addr, remote_addr, elem_sz, upcxx::operation_cx::as_promise(p));
+
+            all_puts[t.rank].first.push_back(local_addr);
+            all_puts[t.rank].second.push_back(remote_addr);
```
Comment (bonachea) [inline src/tamm/tensor_impl.hpp:1162]: @jcorbino Thanks for making this improvement to enable aggregation, it should make an enormous difference to the OFF-node RMA performance.

With that in place, I'd suggest one additional refinement, to reduce overhead for the ON-node case:

```suggestion
            if (remote_addr.is_local()) { // on-node : utilize lightweight shared-memory bypass
                *((T*)remote_addr.local()) = *(T*)local_addr;
            } else { // off-node : aggregated RMA using rput_regular below
                all_puts[t.rank].first.push_back(local_addr);
                all_puts[t.rank].second.push_back(remote_addr);
            }
```

This won't change the network behavior at all, but it should reduce the overhead of `std::vector` manipulation for on-node targets that don't benefit from aggregation (because they are satisfied through shared-memory bypass directly to the memory system, instead of network packets). In particular, with this change single-node runs will automatically skip the vector manipulation path entirely and use very lightweight load/store instructions.

Of course a symmetric refinement can also be made to the `get_raw` function.

**Review by bonachea — CHANGES_REQUESTED** [review]
New commit looks like a great improvement!

I made one additional minor suggestion for further refinement that should reduce overhead for accesses to tiles on peer processes sharing the same compute node.

File: src/tamm/tensor_impl.hpp (hunk 3 above)
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp]: Thanks @bonachea , it certainly looks way better now! I completely forgot to optimize for shared-memory bypass.

**Review by bonachea — APPROVED** [review]
Thanks for all the improvements - looks much better now!

File: src/tamm/tamm_utils.hpp [hunk 4]
```diff
@@ -3009,13 +3004,9 @@ Tensor<TensorType> tensor_block(Tensor<TensorType> tensor, std::vector<int64_t>
   GA_Pgroup_set_default(ga_pg_default);
 #endif
 
-  Tensor<TensorType> pbtensor = btensor;
-  if(!permute.empty()) {
-    pbtensor = permute_tensor<TensorType>(btensor, permute);
-    Tensor<TensorType>::deallocate(btensor);
-  }
+  if(!permute.empty()) btensor = permute_tensor<TensorType>(btensor, permute);
```
Comment (ajaypanyala) [inline src/tamm/tamm_utils.hpp:3007]: @jcorbino-lbl This line needs to be reverted to the previous code. `permute_tensor` returns a new tensor. Assigning it to the existing handle `btensor` results in a leak since the original tensor to which the handle `btensor` was pointing to was never deallocated.

**Review by ajaypanyala — APPROVED** [review]
